doc/man/tinyleaf.8
.\" Automatically generated by Pod::Man 4.07 (Pod::Simple 3.32)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "TINYLEAF 8"
.TH TINYLEAF 8 "2015-09-12" "INN 2.6.3" "InterNetNews Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tinyleaf \- Very simple IHAVE\-only NNTP server
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBtinyleaf\fR \fIspool\fR [\fIprocessor\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtinyleaf\fR is intended to be the simplest possible transit news server
that still does something useful.  It must be run under \fIinetd\fR\|(8) or some
equivalent, and only implements three commands (\s-1HELP, IHAVE,\s0 and \s-1QUIT\s0).
When it receives an article, it saves it into the directory \fIspool\fR and,
if \fIprocessor\fR is given, passes information about the article to
\&\fIprocessor\fR via a pipe.  The file name of the article will be the \s-1MD5\s0
hash of its message-ID, and if a file by that name already exists,
\&\fBtinyleaf\fR will refuse the article, reporting it as a duplicate.
.PP
If \fIprocessor\fR is given, it should specify the path to a program.  That
program is started when \fBtinyleaf\fR starts, and its current working
directory will be \fIspool\fR.  For each article received by \fBtinyleaf\fR, a
single line will be sent to standard input of \fIprocessor\fR.  That line
will consist of the file name of the received article (relative to
\&\fIspool\fR), a single space, and the message-ID of the received article.
Note that the message-ID will be taken from the argument to the \s-1IHAVE\s0
command and may not match the Message-ID: header in the article.  When
\&\fBtinyleaf\fR shuts down, standard input to \fIprocessor\fR will be closed.
.PP
\&\fBtinyleaf\fR does no syntax verification of received articles whatsoever;
it just stores them and optionally passes them off to \fIprocessor\fR.  It
also never deletes articles; normally, \fIprocessor\fR should do that when
it's finished doing whatever it needs to with the article.
.PP
\&\fBtinyleaf\fR expects \s-1NNTP\s0 commands on standard input and replies on
standard output.  Status information and any error messages are sent to
standard error.  It does no authentication; any authentication must be
done by \fIinetd\fR\|(8) or by a wrapper program.  (One simple authentication
mechanism is to invoke \fBtinyleaf\fR via \fItcpd\fR\|(8) from \s-1TCP\s0 wrappers and use
\&\fI/etc/hosts.allow\fR and \fI/etc/hosts.deny\fR to restrict who can talk to the
server.)
.PP
\&\fBtinyleaf\fR has a (currently hard-coded) maximum message size of 1\ \s-1MB\s0 and a
(similarly hard-coded) timeout of ten minutes for each command or chunk of
article data.
.SH "EXAMPLE"
.IX Header "EXAMPLE"
Suppose that you want to archive news articles on a particular host (like
the \s-1FTP\s0 server for a newsgroup archive) where you don't want the overhead
of running a full-blown news server.  Write a program that reads one line
at a time from standard input and treats everything before the first space
as the filename of a news article to archive.  Each time the program reads
a line, it should archive that file and then delete it, and it should exit
when it gets end of file on standard input.
.PP
Then, add a line like:
.PP
.Vb 2
\&    nntp stream tcp nowait archive /usr/sbin/tcpd \e
\&      <pathbin>/tinyleaf <pathspool>/tinyleaf <pathbin>/archive
.Ve
.PP
(all on one line \-\-\ the backslash and split in this line is just for
readability) where \f(CW\*(C`archive\*(C'\fR is the user that owns the archive,
\&\fI/usr/sbin/tcpd\fR is the path to \fItcpd\fR\|(8), \fIpathbin\fR/tinyleaf is
the path to this program, \fIpathspool\fR/tinyleaf is some scratch directory
that the user \f(CW\*(C`archive\*(C'\fR has write access to, and
\&\fIpathbin\fR/archive is the path to your \fBarchive\fR script.
.PP
You can now restrict access to \fBtinyleaf\fR to just your local news server
with \f(CW\*(C`/etc/hosts.allow\*(C'\fR and \f(CW\*(C`/etc/hosts.deny\*(C'\fR and set up an ordinary
feed from the server to the archive host, just like you would to any other
news server, of only the newsgroup that you want to archive.
.PP
Note that the archiving script should probably perform basic syntax and
validity checks on the input, since \fBtinyleaf\fR doesn't.
.PP
This is the application that motivated the original development of this
program.
.SH "BUGS"
.IX Header "BUGS"
The timeout and maximum message size should really be configurable.
\&\fBtinyleaf\fR should also probably not just respond 500 to every command
other than \s-1HELP, IHAVE,\s0 and \s-1QUIT\s0; there are more useful (and more
expected) error codes that could be returned.
.PP
An option to scan the spool directory for any left-over files and pass
them to the processor when starting up would be useful.
.SH "HISTORY"
.IX Header "HISTORY"
Written by Russ Allbery <eagle@eyrie.org> for InterNetNews.
.PP
\&\f(CW$Id:\fR tinyleaf.pod 9767 2014\-12\-07 21:13:43Z iulius $
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIhosts_access\fR\|(5), \fIinetd\fR\|(8), \fItcpd\fR\|(8).


NEWS
Changes in 2.6.3

    * Fixed the selection of the elliptic curve to use with OpenSSL 1.1.0 or
      later; NIST P-256 was enforced instead of using the most secure curve.

    * A new inn.conf parameter has been added to fine-tune the cipher suites
      to use with TLS 1.3:  the *tlsciphers13* now permits configuring them.
      A separate cipher suite configuration parameter is needed for TLS 1.3
      because TLS 1.3 cipher suites are not compatible with TLS 1.2, and
      vice-versa.  In order to avoid issues where legacy TLS 1.2 cipher
      suite configuration configured in the *tlsciphers* parameter would
      inadvertently disable all TLS 1.3 cipher suites, the inn.conf
      configuration has been separated out.

    * Fixed a regression since INN 2.6.1 that prevented articles with
      internationalized header fields (that is to say encoded in UTF-8) from
      being posted.

    * Support for Python 3 has been added to INN.  Embedded Python filtering
      and authentication hooks for innd and nnrpd can now use version 3.3.0
      or later of the Python interpreter.  In the 2.x series, version 2.3.0
      or later is still supported.

      When configuring INN with the --with-python flag, the "PYTHON"
      environment variable, when set, is used to select the interpreter to
      embed.  Otherwise, it is searched in standard paths.

      In case you change the Python interpreter to embed, make sure that the
      Python scripts you use are written in the expected syntax for that
      version of the Python interpreter.  Notably, buffer objects have been
      replaced with memoryview objects in Python 3, and UTF-8 encoding now
      really matters for string literals (Python 3 uses bytes and Unicode
      objects).

      INN documentation and samples of Python hooks have been updated to
      provide more examples.

    * When a Python or Perl filter hook rejects an article, innd now
      mentions the reason in response to CHECK and TAKETHIS commands. 
      Previously, the reason was given only for the IHAVE command.

    * nnrpd now properly logs the hostname of clients whose connection
      failed owing to an issue during the negotiation of a TLS session or
      high load average.

Changes in 2.6.2

    * A new *syntaxchecks* parameter has been added in inn.conf.  It permits
      controlling the level of checks performed by innd and nnrpd.  Up to
      now, only one check can be enabled/disabled:  when *laxmid* is
      mentioned in the values of this new parameter, INN accepts Message-IDs
      that contain ".." in the left part, as well as Message-IDs with two
      "@" (such Message-IDs would otherwise be considered as syntactically
      invalid).  See the inn.conf(5) man page for more details.

      The check is disabled by default (*no-laxmid*), which corresponds to
      the legacy behaviour of INN 2.6.1 and earlier.

    * Use of the ovdb_server helper server is now the default when using the
      ovdb overview method, that is to say the default value for the
      *readserver* parameter in ovdb.conf is now set to true.  It improves
      stability and avoids deadlocks, timing issues and corrupted ovdb
      databases.

    * mailpost now removes empty header fields before attempting to post
      articles, and keeps trace of them in the X-Mailpost-Empty-Hdrs: newly
      generated header field body.  Also, mailpost now sanitizes header
      fields with regards to empty continuation header lines.  Thanks to
      Kamil Jonca for these bug reports.

    * A new -z parameter has been added to mailpost to mention a list of
      header fields to remove from the gated message.  Thanks to Dieter
      Stussy for the patch.

    * Fixed a bug in inews that was rejecting articles containing header
      fields whose length exceeded 998 bytes.  This limitation is for the
      length of a single line of a header field (and not for the length of
      the whole header field, as it was wrongly the case).

    * Added support for GnuPG's gpg binary (in addition to gpgv) in
      pgpverify.  Indeed, gpg still validates signatures made with weak
      digest algorithms like MD5 whereas gpgv no longer do.  Thanks to
      Thomas Hochstein for the patch, which permits validating control
      articles for hierarchies that are still using old PGP keys.

    * Added similar support for GnuPG's gpg binary in perl-nocem to validate
      NoCeM notices from issuers who are still using old PGP keys.

    * A few commands listed in the "Control commands to INND" section in
      daily Usenet reports were appearing as a mere letter; all of them are
      now properly converted to meaningful words.

    * The *tlsprotocols* parameter in inn.conf now recognizes the "TLSv1.3"
      value (for OpenSSL versions implementing TLS 1.3, that is to say
      starting from OpenSSL 1.1.1).

    * The buffindexed overview method will now hopefully work properly on
      systems with a native page size larger than 16KB.

    * Other minor bug fixes and documentation improvements.

Changes in 2.6.1

    * nnrpd now uses -0000 as the time zone for Date: and Injection-Date:
      header fields it generates.  It was previously using +0000, wrongly
      systematically indicating a local time zone at Universal Time when
      *localtime* is set to false (which is the default) in readers.conf. 
      The +0000 time zone will now be used only if *localtime* is set to
      true and UTC is really the local time zone of the server.

    * Julien Elie has implemented in nnrpd the new COMPRESS command
      described in draft-murchison-nntp-compress that extends the NNTP
      protocol to allow a connection to be effectively and efficiently
      compressed.  News clients that also support that extension will be
      able to benefit from that bandwidth optimization and improvement in
      speed.  Moreover, using COMPRESS is more secure than TLS-level

innd/tinyleaf.c
/*  $Id: tinyleaf.c 9767 2014-12-07 21:13:43Z iulius $
**
**  An extremely lightweight receive-only NNTP server.
**
**  Copyright 2003, 2004 Russ Allbery <eagle@eyrie.org>
**
**  Permission is hereby granted, free of charge, to any person obtaining a
**  copy of this software and associated documentation files (the "Software"),
**  to deal in the Software without restriction, including without limitation
**  the rights to use, copy, modify, merge, publish, distribute, sublicense,
**  and/or sell copies of the Software, and to permit persons to whom the
**  Software is furnished to do so, subject to the following conditions:
**
**  The above copyright notice and this permission notice shall be included in
**  all copies or substantial portions of the Software.
**
**  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
**  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
**  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
**  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
**  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
**  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
**  DEALINGS IN THE SOFTWARE.
*/

#include "config.h"
#include "clibrary.h"
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/wait.h>

#include "inn/dispatch.h"
#include "inn/messages.h"
#include "inn/md5.h"
#include "inn/nntp.h"
#include "inn/utility.h"
#include "inn/vector.h"
#include "inn/version.h"
#include "inn/libinn.h"

/* Prototypes for command callbacks. */
static void command_help(struct cvector *, void *);
static void command_ihave(struct cvector *, void *);
static void command_quit(struct cvector *, void *);

/* The actual command dispatch table for TinyNNTP.  This table MUST be
   sorted. */
const struct dispatch commands[] = {
    { "help",  command_help,  0, 0, NULL },
    { "ihave", command_ihave, 1, 1, NULL },
    { "quit",  command_quit,  0, 0, NULL }
};

/* Global state for the daemon. */
struct state {
    struct nntp *nntp;
    unsigned long count;
    unsigned long duplicates;
    FILE *processor;
};


/*
**  Do a clean shutdown, which mostly just involves closing the open processor
**  pipe, if present, and reporting on any abnormal exit status.
*/
static void
shutdown(struct state *state)
{
    int status;

    if (state->processor != NULL) {
        status = pclose(state->processor);
        if (status == -1)
            syswarn("unable to wait for processor");
        else if (WIFEXITED(status) && WEXITSTATUS(status) != 0)
            warn("processor exited with status %d", WEXITSTATUS(status));
        else if (WIFSIGNALED(status))
            warn("processor killed by signal %d", WTERMSIG(status));
    }
    notice("processed %lu articles, rejected %lu duplicates", state->count,
           state->duplicates);
    nntp_free(state->nntp);
    exit(0);
}


/*
**  Handle an IHAVE command.  For the time being, we just ignore the message
**  ID argument.
*/
static void
command_ihave(struct cvector *command, void *cookie)
{
    struct state *state = cookie;
    enum nntp_status status;
    char filename[33], *article, *msgid;
    unsigned char hash[16];
    size_t length;
    int fd, oerrno;

    msgid = xstrdup(command->strings[1]);
    if (!nntp_respond(state->nntp, NNTP_CONT_IHAVE, "Send article"))
        sysdie("cannot flush output");
    status = nntp_read_multiline(state->nntp, &article, &length);
    switch (status) {
    case NNTP_READ_OK:
        break;
    case NNTP_READ_EOF:
        die("connection closed while receiving article");
        shutdown(state);
        break;
    case NNTP_READ_ERROR:
        sysdie("network error while receiving article");
        break;
    case NNTP_READ_TIMEOUT:
        die("network timeout while receiving article");
        break;
    case NNTP_READ_LONG:
        warn("article %s exceeds maximum size", msgid);
        free(msgid);
        return;
    default:
        sysdie("internal: unknown NNTP library status");
        break;
    }
    md5_hash((unsigned char *) msgid, strlen(msgid), hash);
    inn_encode_hex(hash, sizeof(hash), filename, sizeof(filename));
    fd = open(filename, O_WRONLY | O_CREAT | O_EXCL, 0666);
    if (fd < 0) {
        if (errno == EEXIST) {
            if (!nntp_respond(state->nntp, NNTP_FAIL_IHAVE_REFUSE,
                              "Duplicate"))
                sysdie("cannot flush output");
            state->duplicates++;
            free(msgid);
            return;
        }
        sysdie("unable to create article file %s", filename);
    }
    if (xwrite(fd, article, length) < 0) {
        oerrno = errno;
        if (unlink(filename) < 0)
            syswarn("cannot clean up failed write to %s, remove by hand",
                filename);
        errno = oerrno;
        sysdie("unable to write article %s to file %s", msgid,
               filename);
    }
    close(fd);
    if (state->processor != NULL) {
        fprintf(state->processor, "%s %s\n", filename, msgid);
        if (fflush(state->processor) == EOF || ferror(state->processor))
            sysdie("unable to flush %s to processor", filename);
    }
    state->count++;
    if (!nntp_respond(state->nntp, NNTP_OK_IHAVE, "Article received"))
        sysdie("cannot flush output");
    free(msgid);
}


/*
**  Process a HELP command, sending some useful (?) information.
*/
static void
command_help(struct cvector *command UNUSED, void *cookie)
{
    struct state *state = cookie;

    nntp_respond_noflush(state->nntp, NNTP_INFO_HELP, "tinyfeed from %s",
                         INN_VERSION_STRING);
    nntp_send_line_noflush(state->nntp, "Supported commands:");
    nntp_send_line_noflush(state->nntp, " ");
    nntp_send_line_noflush(state->nntp, "  IHAVE <message-id>");
    nntp_send_line_noflush(state->nntp, "  HELP");
    nntp_send_line_noflush(state->nntp, "  QUIT");
    nntp_send_line_noflush(state->nntp, ".");
    if (!nntp_flush(state->nntp))
        sysdie("cannot flush output");
}


/*
**  Process a QUIT command, closing down the server.
*/
static void
command_quit(struct cvector *command UNUSED, void *cookie)
{
    struct state *state = cookie;

    if (!nntp_respond(state->nntp, NNTP_OK_QUIT, "Goodbye"))
        syswarn("cannot flush output during QUIT");
    shutdown(state);
}


/*
**  Process a syntax error.
*/
static void
command_syntax(struct cvector *command UNUSED, void *cookie)
{
    struct state *state = cookie;

    if (!nntp_respond(state->nntp, NNTP_ERR_SYNTAX, "Syntax error"))
        sysdie("cannot flush output");
}


/*
**  Process an unknown command.
*/
static void
command_unknown(struct cvector *command UNUSED, void *cookie)
{
    struct state *state = cookie;

    if (!nntp_respond(state->nntp, NNTP_ERR_COMMAND, "Unsupported command"))
        sysdie("cannot flush output");
}


int
main(int argc, char *argv[])
{
    struct state state = { NULL, 0, 0, NULL };
    struct cvector *command;
    enum nntp_status status;

    message_program_name = "tinyfeed";
    message_handlers_notice(1, message_log_syslog_info);
    message_handlers_warn(1, message_log_syslog_warning);
    message_handlers_die(1, message_log_syslog_warning);

    /* Change to the spool directory where all articles will be written. */
    if (argc < 2)
        die("no spool directory specified");
    if (chdir(argv[1]) < 0)
        sysdie("cannot change directory to %s", argv[1]);

    /* If a processing command was specified, open a pipe to it. */
    if (argc == 3) {
        state.processor = popen(argv[2], "w");
        if (state.processor == NULL)
            sysdie("cannot open a pipe to %s", argv[2]);
        setvbuf(state.processor, NULL, _IONBF, BUFSIZ);
    }

    /* Go into the main input loop.  The only commands we support, for now,
       are HELP, IHAVE and QUIT. */
    notice("starting");
    state.nntp = nntp_new(STDIN_FILENO, STDOUT_FILENO, 1024 * 1024, 10 * 60);
    nntp_respond(state.nntp, NNTP_OK_BANNER_NOPOST, "tinyfeed ready");
    command = cvector_new();
    while (1) {
        status = nntp_read_command(state.nntp, command);
        switch (status) {
        case NNTP_READ_OK:
            dispatch(command, commands, ARRAY_SIZE(commands), command_unknown,
                     command_syntax, &state);
            break;
        case NNTP_READ_EOF:
            notice("connection closed");
            shutdown(&state);
            break;
        case NNTP_READ_ERROR:
            sysdie("network error");
            break;
        case NNTP_READ_TIMEOUT:
            notice("network timeout");
            shutdown(&state);
            break;
        case NNTP_READ_LONG:
            warn("long line received");
            break;
        default:
            sysdie("internal: unknown NNTP library status");
            break;
        }
    }
}
